pex/pex_builder.py

```diff
@@ -769,22 +776,96 @@ def zip_cache_dir(path):
                     os.path.join(internal_cache, location),
                 )
 
-    def _build_zipapp(
-        self,
-        filename,  # type: str
-        deterministic_timestamp=False,  # type: bool
-        compress=True,  # type: bool
-    ):
-        # type: (...) -> None
+    def _cache_compressed_dists_for_stitching(self):
+        # type: () -> Dict[str, str]
+        """Zip up each fingerprinted dist into an atomic shared cache location."""
+        merge_deps = {}  # type: Dict[str, str]
+        for dist_label, fingerprint in self._pex_info.distributions.items():
+            cached_dist = os.path.join(
+                self._pex_info.pex_root,
+                "stitched_dists",
+                fingerprint,
+                dist_label,
+            )
+            with atomic_directory(os.path.dirname(cached_dist)) as atomic_zip_dir:
```

Comment:
For reference #1431 which was followed up a few days later to `s/--spread/--layout packed/`

Reply:
Unfortunately, yes. The `--layout packed` cache zips (correctly) strip `Layout.DEPS_DIR='.deps'` from every entry:
```bash
> unzip -l test.pex/.deps/tensorflow_gpu-2.5.3-cp38-cp38-manylinux2010_x86_64.whl | head -n10 
Archive:  test.pex/.deps/tensorflow_gpu-2.5.3-cp38-cp38-manylinux2010_x86_64.whl
  Length      Date    Time    Name
---------  ---------- -----   ----
       83  1980-01-01 00:00   .layout.json
        0  1980-01-01 00:00   tensorflow_gpu-2.5.3.dist-info/
        4  1980-01-01 00:00   tensorflow_gpu-2.5.3.dist-info/INSTALLER
    13777  1980-01-01 00:00   tensorflow_gpu-2.5.3.dist-info/LICENSE
     2810  1980-01-01 00:00   tensorflow_gpu-2.5.3.dist-info/METADATA
        0  1980-01-01 00:00   tensorflow_gpu-2.5.3.dist-info/REQUESTED
      111  1980-01-01 00:00   tensorflow_gpu-2.5.3.dist-info/WHEEL
````

while the new ones for `--layout zipapp` have the prefix (which is needed for the zip concatenation to work):
```bash
> unzip -l /home/cosmicexplorer/.pex/stitched_dists/7a4763a35d0824ebb172b00f8d0241ff231404c4b7d97dd5ea870d5afca336a4/tensorflow_gpu-2.5.3-cp38-cp38-manylinux2010_x86_64.whl | head -n10
Archive:  /home/cosmicexplorer/.pex/stitched_dists/7a4763a35d0824ebb172b00f8d0241ff231404c4b7d97dd5ea870d5afca336a4/tensorflow_gpu-2.5.3-cp38-cp38-manylinux2010_x86_64.whl
  Length      Date    Time    Name
---------  ---------- -----   ----
        0  1980-01-01 00:00   .deps/
        0  1980-01-01 00:00   .deps/tensorflow_gpu-2.5.3-cp38-cp38-manylinux2010_x86_64.whl/
       83  1980-01-01 00:00   .deps/tensorflow_gpu-2.5.3-cp38-cp38-manylinux2010_x86_64.whl/.layout.json
        0  1980-01-01 00:00   .deps/tensorflow_gpu-2.5.3-cp38-cp38-manylinux2010_x86_64.whl/tensorflow_gpu-2.5.3.dist-info/
        4  1980-01-01 00:00   .deps/tensorflow_gpu-2.5.3-cp38-cp38-manylinux2010_x86_64.whl/tensorflow_gpu-2.5.3.dist-info/INSTALLER
    13777  1980-01-01 00:00   .deps/tensorflow_gpu-2.5.3-cp38-cp38-manylinux2010_x86_64.whl/tensorflow_gpu-2.5.3.dist-info/LICENSE
     2810  1980-01-01 00:00   .deps/tensorflow_gpu-2.5.3-cp38-cp38-manylinux2010_x86_64.whl/tensorflow_gpu-2.5.3.dist-info/METADATA
```

It seems maybe reasonable to generate both of these entries at once from the `packed_wheels` cache dir, but I really don't want to wait TWICE as long to generate two compressed zips for each dist we haven't cached yet.

It's possible we could avoid duplicating the compression work without using two separate cache directories, but it might get complicated.

In general, I see this `stitched_dists` cache as strictly for speeding up the default `--layout zipapp` experience (which means it might contain e.g. invalid zip data in order to feed into `zip -FF`, and may be removed later), while `packed_wheels` contains a pristine directory to conform to the packed pex interface. Does that seem like a reasonable split?